File: csharp/src/AutomationFramework.Core/Clients/ProductsClient.cs
using System.Globalization;
using System.Net.Http.Json;
using AutomationFramework.Core.Models;
using AutomationFramework.Core.Support;

namespace AutomationFramework.Core.Clients;

/// <summary>
/// Catalogue endpoints.
/// </summary>
/// <remarks>
/// One method per operation. No assertions, no status checks, no business logic — those belong to the
/// test. Query and path values are escaped rather than concatenated raw into a URL.
/// <para>
/// The <see cref="HttpClient"/> is injected rather than read from a static. That keeps the dependency
/// visible, lets a test point one client at a different base address, and means these are genuinely
/// instance methods instead of static helpers wearing a class as a hat.
/// </para>
/// </remarks>
public sealed class ProductsClient(HttpClient http)
{
    private const string Products = "products";

    /// <summary><c>GET /products?page=n</c></summary>
    public async Task<ApiResult<PagedProducts>> ListPageAsync(int page)
    {
        string path = string.Create(CultureInfo.InvariantCulture, $"{Products}?page={page}");
        using HttpResponseMessage response = await http.GetAsync(path).ConfigureAwait(false);
        return await ApiResult.OfAsync<PagedProducts>(response).ConfigureAwait(false);
    }

    /// <summary><c>GET /products/{id}</c></summary>
    public async Task<ApiResult<Product>> GetByIdAsync(string productId)
    {
        string path = $"{Products}/{Uri.EscapeDataString(productId)}";
        using HttpResponseMessage response = await http.GetAsync(path).ConfigureAwait(false);
        return await ApiResult.OfAsync<Product>(response).ConfigureAwait(false);
    }

    /// <summary><c>GET /products/search?q=...</c></summary>
    public async Task<ApiResult<PagedProducts>> SearchAsync(string query)
    {
        string path = $"{Products}/search?q={Uri.EscapeDataString(query)}";
        using HttpResponseMessage response = await http.GetAsync(path).ConfigureAwait(false);
        return await ApiResult.OfAsync<PagedProducts>(response).ConfigureAwait(false);
    }

    /// <summary><c>GET /products?by_brand=&lt;brandId&gt;</c></summary>
    public async Task<ApiResult<PagedProducts>> ListByBrandAsync(string brandId)
    {
        string path = $"{Products}?by_brand={Uri.EscapeDataString(brandId)}";
        using HttpResponseMessage response = await http.GetAsync(path).ConfigureAwait(false);
        return await ApiResult.OfAsync<PagedProducts>(response).ConfigureAwait(false);
    }

    /// <summary>
    /// <c>POST /products</c> — creation requires an admin token.
    /// </summary>
    /// <remarks>
    /// Present so the authorization boundary can be tested: the same call with no token, with a customer
    /// token, and with an admin token must behave differently. The header is nullable so the anonymous
    /// case does not need a separate method.
    /// </remarks>
    public async Task<ApiResult<Product>> CreateAsync(string? bearerHeaderValue, object payload)
    {
        using HttpRequestMessage request = new(HttpMethod.Post, Products)
        {
            Content = JsonContent.Create(payload, options: ApiHttp.JsonOptions),
        };

        if (bearerHeaderValue is not null)
        {
            request.Headers.TryAddWithoutValidation("Authorization", bearerHeaderValue);
        }

        using HttpResponseMessage response = await http.SendAsync(request).ConfigureAwait(false);
        return await ApiResult.OfAsync<Product>(response).ConfigureAwait(false);
    }

    /// <summary><c>DELETE /products/{id}</c> — tolerates an already-deleted record, for use in cleanup.</summary>
    public async Task<ApiResult<NoBody>> DeleteIgnoringMissingAsync(
        string bearerHeaderValue,
        string productId)
    {
        string path = $"{Products}/{Uri.EscapeDataString(productId)}";
        using HttpRequestMessage request = new(HttpMethod.Delete, path);
        request.Headers.TryAddWithoutValidation("Authorization", bearerHeaderValue);

        using HttpResponseMessage response = await http.SendAsync(request).ConfigureAwait(false);
        return await ApiResult.OfStatusOnlyAsync(response).ConfigureAwait(false);
    }
}

File: java/src/main/java/com/company/automation/clients/ProductsClient.java
package com.company.automation.clients;

import static io.restassured.RestAssured.given;

import com.company.automation.models.PagedProducts;
import com.company.automation.models.Product;
import com.company.automation.support.ApiSpec;
import io.restassured.response.Response;

/**
 * Catalogue endpoints.
 *
 * <p>One method per operation. No assertions, no status checks, no business logic — those belong to
 * the test. Path and query values go through REST Assured's parameter handling so they are encoded
 * properly rather than concatenated into a URL.
 */
public final class ProductsClient {

  private static final String PRODUCTS = "/products";

  /** {@code GET /products?page=n} */
  public ApiResult<PagedProducts> listPage(int page) {
    Response response =
        given().spec(ApiSpec.spec()).queryParam("page", page).when().get(PRODUCTS).thenReturn();
    return ApiResult.of(response, PagedProducts.class);
  }

  /** {@code GET /products/{id}} */
  public ApiResult<Product> getById(String productId) {
    Response response =
        given()
            .spec(ApiSpec.spec())
            .pathParam("id", productId)
            .when()
            .get(PRODUCTS + "/{id}")
            .thenReturn();
    return ApiResult.of(response, Product.class);
  }

  /** {@code GET /products/search?q=...} */
  public ApiResult<PagedProducts> search(String query) {
    Response response =
        given()
            .spec(ApiSpec.spec())
            .queryParam("q", query)
            .when()
            .get(PRODUCTS + "/search")
            .thenReturn();
    return ApiResult.of(response, PagedProducts.class);
  }

  /** {@code GET /products?by_brand=<brandId>} */
  public ApiResult<PagedProducts> listByBrand(String brandId) {
    Response response =
        given()
            .spec(ApiSpec.spec())
            .queryParam("by_brand", brandId)
            .when()
            .get(PRODUCTS)
            .thenReturn();
    return ApiResult.of(response, PagedProducts.class);
  }

  /**
   * {@code POST /products} — creation requires an admin token.
   *
   * <p>Present so the authorization boundary can be tested: the same call with no token, with a
   * customer token, and with an admin token must behave differently.
   */
  public ApiResult<Product> create(String bearerHeaderValue, String jsonBody) {
    Response response =
        given()
            .spec(ApiSpec.spec())
            .header("Authorization", bearerHeaderValue)
            .body(jsonBody)
            .when()
            .post(PRODUCTS)
            .thenReturn();
    return ApiResult.of(response, Product.class);
  }

  /** {@code DELETE /products/{id}} — tolerates an already-deleted record, for use in cleanup. */
  public ApiResult<Void> deleteIgnoringMissing(String bearerHeaderValue, String productId) {
    Response response =
        given()
            .spec(ApiSpec.spec())
            .header("Authorization", bearerHeaderValue)
            .pathParam("id", productId)
            .when()
            .delete(PRODUCTS + "/{id}")
            .thenReturn();
    return ApiResult.ofStatusOnly(response);
  }
}
